hw/formal/tools/jaspergold/parse-fpv-report.py

```diff
@@ -0,0 +1,217 @@
+
+    except IOError as err:
+        log.error("[get_cov_results] IOError %s", err)
+        cov_results = None
+
+    return cov_results
+
+
+def main():
+    parser = argparse.ArgumentParser(
+        description=
+        '''This script parses the JasperGold FPV log to extract below information.
+
+        "messages": {
+          "errors"      : []
+          "warnings"    : []
+          "cex"         : ["property1", "property2"...],
+          "undetermined": [],
+          "unreachable" : [],
+        },
+        "fpv_summary": {
+          "errors"      : 0
+          "warnings"    : 2
+          "proven"      : 20,
+          "cex"         : 5,
+          "covered"     : 18,
+          "undetermined": 7,
+          "unreachable" : 2,
+          "pass_rate"   : "90 %",
+          "cover_rate"  : "90 %"
+        },
+        If coverage is enabled, this script will also parse the coverage result:
+        "fpv_coverage": {
+          stimuli: "90 %",
+          coi    : "90 %",
+          proof  : "80 %"
+        }
+        The script returns nonzero status if any errors or property failures including
+        "cex, undetermined, unreachable" are presented.
+        ''')
+    parser.add_argument('--logpath',
+                        type=str,
+                        default="fpv.log",
+                        help=('FPV log file path. Defaults to `fpv.log` '
+                              'under the current script directory.'))
+
+    parser.add_argument('--reppath',
+                        type=str,
+                        default="results.hjson",
+                        help=('Parsed output hjson file path. Defaults to '
+                              '`results.hjson` under the current script directory.'))
+
+    parser.add_argument('--cov',
+                        type=int,
+                        default=0,
+                        help=('Enable parsing coverage data. '
+                              'Default coverage parsing is disabled.'))
+
+    args = parser.parse_args()
+
+    results = get_results(args.logpath)
+
+    if args.cov:
+        results["fpv_coverage"] = get_cov_results(args.logpath)
+
+    with Path(args.reppath).open("w") as results_file:
+        hjson.dump(results,
+                   results_file,
+                   ensure_ascii=False,
+                   for_json=True,
+                   use_decimal=True)
+
+    # return nonzero status if any errors or property failures are present
+    # TODO: currently allow FPV warnings
+    err_msgs = results["messages"]
+    n_errors = len(err_msgs["errors"])
+    n_failures = (len(err_msgs.get("cex")) + len(err_msgs.get("undetermined")) +
+                  len(err_msgs.get("unreachable")))
+    if n_errors > 0 or n_failures > 0:
+        log.info("FPV error %d, FPV property failures %d", n_errors, n_failures)
+        sys.exit(1)
+    else:
+        log.info("FPV logfile parsed succesfully")
+        sys.exit(0)
+
+
+if __name__ == "__main__":
+    main()
```

Comment:
I think it's probably neater just to return from main (like you'd do in C) and put the call to `sys.exit` at top-level. Like this:

```python
    ...
    if n_errors > 0 or n_failures > 0:
        log.info("FPV error %d, FPV property failures %d", n_errors, n_failures)
        return 1

    log.info("FPV logfile parsed succesfully")
    return 0


if __name__ == "__main__":
    sys.exit(main())
```

Reply:
Oh, and maybe "FPV errors" (with the 's'), otherwise "FPV error 13" sounds like an error code, rather than a count.